src/connection/mod.rs

```diff
@@ -0,0 +1,160 @@
+    where
+        S: PropertyStore,
+    {
+        let device_properties = store.device_props().await?;
+        let server_interfaces = Self::filter_server_interfaces(interfaces);
+
+        Ok(Self {
+            interfaces: interfaces.get_introspection_string(),
+            server_interfaces,
+            device_properties,
+        })
+    }
+}
+
+/// Holds generic event data such as interface name and path
+/// The payload must be deserialized after verification with the
+/// specific [`Connection::deserialize_individual`] or [`Connection::serialize_individual`]
+pub(crate) struct ReceivedEvent<P> {
+    pub(crate) interface: String,
+    pub(crate) path: String,
+    pub(crate) payload: P,
+}
+
+#[async_trait]
+pub(crate) trait Connection<S>: Clone {
+    type Payload: Send + Sync + 'static;
+
+    /// Method called imediately after the Connection gets created
+    async fn connect(&self, introspection: Introspection) -> Result<(), crate::Error>
+    where
+        S: PropertyStore;
+
+    /// This function returns the next event from the connection
+    /// and waits for it as necessary. It's important to note that not
+    /// every received incoming event must get returned from this method.
+    /// Implementations could decide to process internally some types of
+    /// incoming messages.
+    async fn next_event(
+        &self,
+        device: &SharedDevice<S>,
+    ) -> Result<ReceivedEvent<Self::Payload>, crate::Error>
+    where
+        S: PropertyStore;
+
+    /// Deserializes a received payload to an individual astarte value
+    fn deserialize_individual(
+        &self,
+        mappig: MappingRef<'_, &Interface>,
+        payload: &Self::Payload,
+    ) -> Result<(AstarteType, Option<Timestamp>), crate::Error>;
+
+    /// Deserializes a received payload to an aggregate object
+    fn deserialize_object(
+        &self,
+        object: ObjectRef,
+        path: &MappingPath<'_>,
+        payload: &Self::Payload,
+    ) -> Result<(HashMap<String, AstarteType>, Option<Timestamp>), crate::Error>;
+
+    /// Sends individual [`AstarteType`] values over this connection
+    async fn send_individual(&self, data: ValidatedIndividual<'_>) -> Result<(), crate::Error>
+    where
+        S: 'static;
+
+    /// Sends objects [`HashMap`] values over this connection
+    async fn send_object(&self, data: ValidatedObject<'_>) -> Result<(), crate::Error>
+    where
+        S: 'static;
+}
+
+#[async_trait]
+pub(crate) trait Registry {
+    /// Method to start receiving events related to the passed interface
+    async fn subscribe(&self, interface: &str) -> Result<(), crate::Error>;
+
+    /// Method that stops the connection from receiving events related to the passed interface
+    async fn unsubscribe(&self, interface: &str) -> Result<(), crate::Error>;
+
+    /// Method that sends the introspection [`String`] over the connection
+    async fn send_introspection(&self, introspection: String) -> Result<(), crate::Error>;
```

Comment:
The introspection is a concept related to Astarte MQTT Transport, so there is no need to expose it in a trait.

Reply:
As stated above, this trait has been removed while adding some methods to the Connection trait